nutils/numeric.py

```diff
@@ -24,6 +24,70 @@
 
 from . import types, warnings
 import numpy, numbers, builtins, collections.abc
+from ctypes import byref, c_int, c_ssize_t, c_void_p, c_char_p, py_object, pythonapi, Structure, POINTER
+c_ssize_p = POINTER(c_ssize_t)
+
+class Py_buffer(Structure):
+  '''Low level access to Python buffers.
+
+  The buffer structure exposes information about the memory the underlies bytes
+  objects, Numpy arrays, and other objects that implement Python's buffer
+  protocol. This is useful for instance to establish if two separate objects
+  provide views into the same memory location, which can be a cheap alternative
+  to a full data comparison. To support this use case, the Py_buffer object is
+  hasheable and support equality testing.
+  '''
+
+  # Unfortunately, the buffer protocol does not have a Python facing API. The
+  # closest thing is memoryview, which exposes most of Py_buffer but not the
+  # actual address. While we could abuse Numpy's array interface, this turns
+  # out to be slower than the ctypes route even if obj already is an ndarray.
+
+  __slots__ = ()
+
+  # Structure definition: https://docs.python.org/3/c-api/buffer.html
+  _fields_ = [
+    ('buf', c_void_p),
+    ('obj', py_object),
+    ('len', c_ssize_t),
+    ('itemsize', c_ssize_t),
+    ('readonly', c_int),
+    ('ndim', c_int),
+    ('format', c_char_p),
+    ('shape', c_ssize_p),
+    ('strides', c_ssize_p),
+    ('suboffsets', c_ssize_p),
+    ('internal', c_void_p)]
+
+  # Constant definitions: https://github.com/python/cpython/blob/master/Include/cpython/object.h
+  _flags = c_int(0x1c) # PyBUF_STRIDES | PyBUF_ND | PyBUF_FORMAT
+
+  def __init__(self, obj):
+    pythonapi.PyObject_GetBuffer(py_object(obj), byref(self), self._flags)
+
+  def __hash__(self):
+    return hash((self.buf, self.len, self.ndim))
+
+  def __eq__(self, other):
+    return isinstance(other, Py_buffer) \
+      and self.buf == other.buf and self.ndim == other.ndim and self.format == other.format \
+      and all(self.shape[i] == other.shape[i] and self.strides[i] == other.strides[i] for i in range(self.ndim))
+
+  def __del__(self):
+    pythonapi.PyBuffer_Release(byref(self))
+
+  def replace_obj(self, obj):
+    '''Replace the obj attribute (dangerous!)
+
+    The buffer's ``obj`` attribute serves to keep the object alive that owns
+    the memory block. This method may be used to swap the object by another
+    that serves the same purpose. Note that if the new object fails to keep the
+    memory alive this may lead to fatal errors.
+    '''
+
+    pythonapi.Py_DecRef(py_object(self.obj))
+    pythonapi.Py_IncRef(py_object(obj))
```

Comment:
Should we increase the reference count before decrease in case `obj` and `self.obj` refer to the same thing?

Reply:
We could, but since the function frame has its own reference it should be impossible that `obj` gets released between these two lines.